src/frontend/qt_sdl/LuaMain.cpp

```diff
@@ -0,0 +1,573 @@
+    event->accept();
+}
+
+void LuaConsoleDialog::onOpenScript()
+{
+    QFileInfo file = QFileInfo(QFileDialog::getOpenFileName(this, "Load Lua Script",QDir::currentPath()));
+    if (!file.exists())
+        return;
+    currentScript=file;
+    LuaScript::FlagNewLua = true;
+}
+
+LuaConsole::LuaConsole(QWidget* parent)
+{
+    this->setParent(parent);
+}
+
+void LuaConsole::onGetText(QString string)
+{
+    this->appendPlainText(string);
+    QScrollBar* bar = verticalScrollBar();
+    bar->setValue(bar->maximum());
+}
+
+void EmuThread::onLuaPrint(QString string)
+{
+    emit signalLuaPrint(string);
+}
+
+void LuaConsole::onClear()
+{
+    this->clear();
+}
+
+void EmuThread::onLuaClearConsole()
+{
+    emit signalLuaClearConsole();
+}
+
+
+void LuaScript::luaClearConsole()
+{
+    LuaDialog->console->clear();
+}
+
+void EmuThread::onLuaSaveState(QString string)
+{
+    emit signalLuaSaveState(string);
+}
+
+void EmuThread::onLuaLoadState(QString string)
+{
+    emit signalLuaLoadState(string);
+}
+
+
+
+std::vector<LuaScript::LuaFunction*> LuaFunctionList; // List of all lua functions.
+QWidget* LuaScript::panel=nullptr;
+lua_State* LuaScript::MainLuaState=nullptr;
+volatile bool LuaScript::FlagPause=false;
+volatile bool LuaScript::FlagStop=false;
+volatile bool LuaScript::FlagNewLua=false;
+
+LuaScript::LuaFunction::LuaFunction(LuaScript::luaFunctionPointer cf,const char* n,std::vector<LuaScript::LuaFunction*>* container)
+{
+    this->cfunction=cf;
+    this->name=n;
+    container->push_back(this);
+}
+
+
+#define MELON_LUA_HOOK_INSTRUCTION_COUNT 50 //number of vm instructions between hook calls
+void LuaScript::luaHookFunction(lua_State* L, lua_Debug *arg)
+{
+    if(FlagStop and (arg->event == LUA_HOOKCOUNT))
+        luaL_error(L, "Force Stopped");
+}
+
+void LuaScript::createLuaState()
+{
+    if(!FlagNewLua)
+        return;
+    LuaScript::LuaOverlays.clear();
+    FlagNewLua = false;
+    MainLuaState=nullptr;
+    std::string fileName = LuaDialog->currentScript.fileName().toStdString();
+    std::string filedir = LuaDialog->currentScript.dir().path().toStdString();
+    lua_State* L = luaL_newstate();
+    luaL_openlibs(L);
+    for(LuaFunction* function : LuaFunctionList)
+        lua_register(L,function->name,function->cfunction);
+    std::filesystem::current_path(filedir.c_str());
+    lua_sethook(L,&LuaScript::luaHookFunction,LUA_MASKCOUNT,MELON_LUA_HOOK_INSTRUCTION_COUNT);
+    FlagStop = false;
+    if (luaL_dofile(L,&fileName[0])==LUA_OK)
+    {
+        MainLuaState=L;
+        FlagPause=false;
+    }
+    else //Error loading script
+    {
+        emuThread->onLuaPrint(lua_tostring(L,-1));
+        MainLuaState=nullptr;
+
+    }
+}
+
+void LuaConsoleDialog::onStop()
+{
+    LuaScript::FlagStop = true;
+}
+
+void LuaConsoleDialog::onPausePlay()
+{
+    LuaScript::FlagPause = !LuaScript::FlagPause;
+}
+
+//Gets Called once a frame
+void LuaScript::luaUpdate()
+{
+    if(!MainLuaState or FlagPause)
```

Comment:
using or for || is valid C++?

Reply:
Yes
https://en.cppreference.com/w/cpp/language/operator_alternative
Since C++
You can also use it in C by including iso646.h, but it's just some macros